dev/ProgressRing/ProgressRing.cpp

```diff
@@ -84,29 +121,70 @@ void ProgressRing::OnBackgroundPropertyChanged(const winrt::DependencyObject&, c
 
 void ProgressRing::OnBackgroundColorPropertyChanged(const winrt::DependencyObject&, const winrt::DependencyProperty&)
 {
-    if (auto&& player = m_player.get())
+    if (auto&& determinatePlayer = m_determinatePlayer.get())
+    {
+        if (auto const progressRingDeterminate = determinatePlayer.Source())
+        {
+            SetLottieBackgroundColor(progressRingDeterminate);
+        }
+    }
+
+    if (auto&& indeterminatePlayer = m_indeterminatePlayer.get())
     {
-        if (auto const progressRingIndeterminate = player.Source().try_as<AnimatedVisuals::ProgressRingIndeterminate>())
+        if (auto const progressRingIndeterminate = indeterminatePlayer.Source())
         {
             SetLottieBackgroundColor(progressRingIndeterminate);
         }
     }
 }
 
+void ProgressRing::OnOpacityPropertyChanged(const winrt::DependencyObject&, const winrt::DependencyProperty&)
+{
+    if (auto&& indeterminatePlayer = m_indeterminatePlayer.get())
+    {
+        auto test = indeterminatePlayer.Opacity();
+
+        if (indeterminatePlayer.Opacity() == 0)
+        {
+            indeterminatePlayer.Stop();
+        }
+    }
+}
+
 void ProgressRing::OnIsActivePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
 {
-    ChangeVisualState();
+    UpdateStates();
+}
+
+void ProgressRing::OnIsIndeterminatePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
 }
 
 void ProgressRing::SetAnimatedVisualPlayerSource()
 {
-    if (auto&& player = m_player.get())
+    if (auto&& determinatePlayer = m_determinatePlayer.get())
+    {
+        if (!determinatePlayer.Source())
```

Comment:
>determinatePlayer [](start = 13, length = 17)

how could this be null?

Reply:
It will return null if there is no custom LottieSource set with re-templating.